KAFKA-5494: enable idempotence with max.in.flight.requests.per.connection > 1

apurvam opened the pull request:
Here we introduce client and broker changes to support multiple inflight requests while still guaranteeing idempotence. Two major problems to be solved:

1. Sequence number management on the client when there are request failures. When a batch fails,  future inflight batches will also fail with `OutOfOrderSequenceException`. This must be handled on the client with intelligent sequence reassignment. We must also deal with the fatal failure of some batch: the future batches must get different sequence numbers when the come back.
2. On the broker, when we have multiple inflights, we can get duplicates of multiple old batches. With this patch, we retain the record metadata for 5 older batches. 

I have added `TODO(reviewers)` comments for specific decisions in the code which are worth discussing.

TODO: 
1. Add more unit tests, especially for loading different snapshot versions correctly, more client side unit tests, more broker side tests to validate that we are caching the correct number of batches (some of this is already there).
2. Update the system tests to check for ordering. 
3. Run a tight loop of system tests. 
4. Add comments about the assumptions made around the network client semantics of send/receive.
5. Gracefully handle changes in the producer id. In particular, we should never change the producer id once a batch has been sent, only the sequence number. This protects against duplicates and is a generalization of the previous approach to not reset producer state on retry.

Commit:
Initial commit of client side changes with some tests

Commit:
Implemented broker side changes to cache extra metadata.

Todo:
  1) Write more unit tests.
  2) Handle deletion / retention / cleaning correctly.

Commit:
Change the client side code so that the sequence numbers are assigned
and incremented during drain. If a batch is retried, it's sequence
number is unset during the completion handler. If the first inflight
batch returns an error, the next sequence to assign is reset to the last
ack'd sequence + 1.

Commit:
WIP

Commit:
Implemented log cleaning functionality with tests

Commit:
Fix merge issues aftre rebasing onto trunk

apurvam commented:
cc @ijuma @hachikuji

Commit:
Remove some unneeded variables and function parameters

Commit:
More minor cleanups

ijuma commented on core/src/test/resources/log4j.properties:
```diff
@@ -19,7 +19,6 @@ log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
 log4j.appender.stdout.layout.ConversionPattern=[%d] %p %m (%c:%L)%n
 
 log4j.logger.kafka=ERROR
-log4j.logger.org.apache.kafka=ERROR
```
Probably unintentional.

apurvam replied:
Yes. I will fix the log4j.properties as we near the end of the review process.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -476,18 +476,10 @@ private static int configureRetries(ProducerConfig config, boolean idempotenceEn
+    private static int configureInflightRequests(ProducerConfig config, boolean idempotenceEnabled) {
+        if (idempotenceEnabled && 5 < config.getInt(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION)) {
+            throw new ConfigException("Must set " + ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION + " to at most 5" +
+                    " to use the idempotent producer. Otherwise we cannot guarantee idempotence.");
```
Instead of saying we can't guarantee idempotence, maybe we can just say that idempotence is only supported for in-flight requests greater than 5. This makes it sound like it's a correctness concern, but it's not really.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -235,7 +235,7 @@ public RecordAppendResult append(TopicPartition tp,
     }
 
-    private MemoryRecordsBuilder recordsBuilder(ByteBuffer buffer, byte maxUsableMagic) {
+    private MemoryRecordsBuilder recordsBuilder(TopicPartition topicPartition, ByteBuffer buffer, byte maxUsableMagic) {
```
nit: seems we don't use `topicPartition`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -472,17 +473,25 @@ public boolean hasUndrained() {
+                                            batch.setProducerState(producerIdAndEpoch, transactionManager.sequenceNumber(batch.topicPartition), isTransactional);
+                                            transactionManager.incrementSequenceNumber(batch.topicPartition, batch.recordCount);
+                                            log.debug("Assigned producerId {} and producerEpoch {} to batch with sequence " +
+                                                            "{} being sent to partition {}", producerIdAndEpoch.producerId,
```
nit: might help if the log message metions that this is the base sequence.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -68,6 +68,7 @@
     public final String logPrefix;
 
     private final Map<TopicPartition, Integer> sequenceNumbers;
+    private final Map<TopicPartition, Integer> lastAckedSequence;
```
nit: would be nice to have consistent naming (e.g. `sequences` and `lastAckedSequences`). Also, a comment which clarifies what purpose each serves (especially for the former).

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -472,17 +473,25 @@ public boolean hasUndrained() {
 
                                         ProducerBatch batch = deque.pollFirst();
-                                        if (producerIdAndEpoch != null && !batch.inRetry()) {
+                                        if (producerIdAndEpoch != null) {
```
Seems we need to update the comment below with this check gone.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -425,6 +425,7 @@ private void maybeWaitForProducerId() {
                         ProducerIdAndEpoch producerIdAndEpoch = new ProducerIdAndEpoch(
                                 initProducerIdResponse.producerId(), initProducerIdResponse.epoch());
                         transactionManager.setProducerIdAndEpoch(producerIdAndEpoch);
+                        return;
```
Wonder if we can improve the structure to make this kind of bug less likely. Instead of falling through to hit the backoff case, maybe we could just move the retry logic into the cases where it actually applies?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -520,7 +524,7 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
+                            batch.baseSequence());
                     reenqueueBatch(batch, now);
                 } else {
                     failBatch(batch, response, new OutOfOrderSequenceException("Attempted to retry sending a " +
```
Seems like we always call `recordErrors` after failing a batch. Maybe we could just move that call into `failBatch`?

hachikuji replied:
My understanding is that we reset the producerId if we find one partition for which we have expired all in-flight requests. My question is how this affects the in-flight requests of other partitions? If one of them happens to need retry, wouldn't we hit this case? But since the sequence number isn't changing, I think it should be safe to continue retrying with the old producerId.

Alternatively, maybe we should drain all partitions before resetting the producerId. That's a bit heavy-handed, but hopefully this case is rare in practice.

apurvam replied:
This is a good point. Yes resetting the producer id due to one bad partition will affect in flight batches of other partitions which happen to need a retry. Our original take on this was 'bad luck, once you have expiring batches, this is just an additional bit of funkiness'. Personally, I don't think we need to over optimize this right now. If it becomes a usability issue, we can fix it in the future.

apurvam replied:
I filed a JIRA to track this: https://issues.apache.org/jira/browse/KAFKA-5870

I think we should solve this --existing-- bug in 1.0.0, but I would like to do it after the code for KIP-192 goes in.

hachikuji replied:
Sounds good.

apurvam commented:
Here are my todos for the patch, to help reviewers.

    1. Snapshot load unit tests.
    2. ProducerState tests to ensure that the correct number of batches are retained when truncation happens at head. 
    3. Log tests to ensure that we handle duplicates of batches older than the last appended batch.
    4. Review client tests. Do we need to add anything around batch expiry? We probably should add assertions around resetting the batch metadata on retry if we don’t do it already.
    5. How does KIP-91 discussion impact this? We probably should retain the producer id and only set a new sequence if the producer id hasn’t changed. This is a generalization of the previous `isRetry` check, where we don’t change the producer id of a batch once assigned.
    6. Need to upgrade the system test to enforce ordering.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -495,6 +496,9 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
                                long now) {
         Errors error = response.error;
+        if (error != Errors.NONE)
+            maybeResetNextSequenceForPartition(batch.topicPartition, batch.baseSequence());
```
nit: would be nice if the name of this function was somehow indicative of the expectation that it is only for a failure case. Something like `maybeResetSequenceOnFailure`?

hachikuji commented on clients/src/test/resources/log4j.properties:
```diff
@@ -19,3 +19,5 @@ log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
 
 log4j.logger.org.apache.kafka=ERROR
+
+log4j.logger.org.apache.kafka.clients.producer.internals=TRACE
```
Inadvertent I assume.

apurvam replied:
Yes. I will leave this in here until just before the final commit.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -765,18 +765,25 @@ class Log(@volatile var dir: File,
+      // if this is a client produce request, there will be upto 5 batches which could have been duplicated.
+      // If we find a duplicate, we return the metadata of the appended batch to the client.
+      if (isFromClient) {
+        maybeLastEntry match {
```
I think this works:
```scala
        maybeLastEntry.map(_.duplicateOf(batch)).foreach { duplicate =>
          return (updatedProducers, completedTxns.toList, duplicate)
        }
```

hachikuji replied:
Forgot to make this change?

apurvam replied:
Actually, I did try your suggestion, but it didn't quite work: for some reason spurious duplicates were being detected for some reason. So I left the old code as is.

hachikuji replied:
Ah, a small bug in my snippet. Needed to use `flatMap`:
```scala
        maybeLastEntry.flatMap(_.duplicateOf(batch)).foreach { duplicate =>
          return (updatedProducers, completedTxns.toList, Some(duplicate))
        }
```

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,94 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+  }
+}
+
+private[log] case class ProducerIdEntry(producerId: Long, batchMetadata: mutable.Queue[BatchMetadata],
```
A bit unfortunate that we had to make this class mutable. If it can't be helped, we should make it a normal non-case class. Also, some brief docs on the `batchMetadata` queue would be helpful (in particular in regard to ordering).

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,94 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+ }
+
+  def duplicateOf(batch: RecordBatch): Option[BatchMetadata] = {
+    if (batch.producerEpoch() != producerEpoch)
```
nit: no need for parenthesis for non-mutating getters like `producerEpoch`. A few more of these below (`baseSequence` and `lastSequence`).

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,94 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+  }
+
+  def removeBatchesOlderThan(offset: Long) = {
+    val retainedMetadata = batchMetadata.filter(_.lastOffset >= offset)
```
I'm assuming offsets increase monotonically, so could this be
```scala
batchMetadata.dropWhile(_.lastOffset < offset)
```

hachikuji commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -513,9 +513,9 @@ private[log] class Cleaner(val id: Int,
+        // check if the batch metadata is currently being cached. if so, we cannot
         // remove the batch just yet or the producer may see an out of sequence error.
-        if (batch.hasProducerId && activeProducers.get(batch.producerId).exists(_.lastSeq == batch.lastSequence))
+        if (batch.hasProducerId && activeProducers.get(batch.producerId).exists(_.hasBatchWithSequenceRange(batch.baseSequence, batch.lastSequence).nonEmpty))
```
Hmm.. Now `ProducerIdEntry` is mutable, so it seems like we need a lock. Alternatively we could create an immutable `ProducerIdSummary` which maybe just gives us the sequence range.

I think the comment needs to be updated as well. We only need to maintain the last batch to avoid the out of sequence error; the other batches are only needed to rebuild the cache if we have to. 

I'm wondering how critical it actually is to retain these entries. Suppose for the sake of argument that we do not bother retaining the last 5 entries for a producer; we just keep the last one like now. These entries can still be cached in memory which means they will be available on the leader and follower for duplicate detection after being written, but they may be "lost" upon recovery. My question is basically how likely it is that we will still need the entries after recovering the log? 

I think we would have to be in a situation where 1) the producer has a bunch of in-flights pending, 2) the broker successfully applies several of them and crashes before returning responses, and 3) the broker is restarted, recovers the log, and is elected leader before the producer has gotten all responses. Since we know we have to return Duplicate in some cases anyway, maybe this case is rare enough that we could do the same here?

apurvam replied:
I agree with you that it is a bit overkill to retain 5 batches in the producer snapshot and also in the log. The cases where we would actually have to rely on them seem quite impossible: you need to a broker to fail, the new leader to be elected, and the leadership to come back to the original broker while the client is still retrying. Alternately, all the brokers may bounce together and lose their in memory state. 

In either case, I think it is fine to return a `DuplicateSequenceException` if we don't have the record metadata. This would mean that the broker side changes are minimal. If we want to cover this case, we can always do so in the future: it is just an optimization. 

I will go ahead and backout the changes to the producer snapshot and retention logic, and add the code to raise a duplicate sequence exception if we he are trying to append a batch with an older sequence but don't have the record metadata in memory.

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,94 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+ }
+
+  // Return the batch metadata of the cached batch having the exact sequence range, if any.
+  def hasBatchWithSequenceRange(firstSeq: Int, lastSeq:Int) : Option[BatchMetadata] = {
```
nit: add space after `lastSeq:` and remove the space after the parenthesis. Also, with a name like `hasXX`, I would expect this to return a boolean. Maybe the name should just be `batchWithSequenceInRange`?

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,94 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+    }
+
+    if (duplicate.size > 1)
+      throw new IllegalStateException(s"Found ${duplicate.size} duplicate batches in the cached producer metadata " +
```
Not really sure this is worth checking here. Shouldn't we have ensured that duplicates weren't added to `batchMetadata` in the first place.

apurvam replied:
Yes. This definitely seems spurious. I was debating not including this check, and your comment confirms that I should remove it.

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,94 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+        s"for producerId: $producerId and producerEpoch: $producerEpoch. This means that it's likely there are " +
+        s"duplicates in the log as well.")
+
+    if (duplicate.size == 1)
```
You can replace this with `duplicate.headOption`.

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -187,44 +239,35 @@ private[log] class ProducerAppendInfo(val producerId: Long,
-      // of size 1 which uses the last appended sequence number.
-      this.firstSeq = this.lastSeq
-    }
+    // TODO(reviewers): The semantics of the ProducerIdEntry have changed so that now explicitly caches the metadata
```
Hmm.. It's not clear that you're handling the case that the marker has an epoch bump below. The old sequence numbers will not be valid for the new epoch, so we should clear them as the comment below says? 

As for the last offset being potentially out of sync, I'm not sure it's a problem, but it makes me a tad uncomfortable. We use this to determine when a producer should be expired (due to retention enforcement), but it may be OK if we expire a producerId when the only record left is the marker. However, we might need to be a little careful when loading the producer state from the log to make sure that producer expiration is determined consistently. Maybe we should at least rename the `lastOffset` method to `lastDataOffset` or something like that to emphasize that it does not take control records into account.

apurvam replied:
Actually, the case where the control batch has a new epoch is being handled in `currentEntry.maybeUpdateEpoch`. If the epoch changes we update the epoch in the `ProducerIdEntry` and clear all the batch metadata, which resets all the sequence numbers producer id. 

I am not sure I follow your point about producer id expiration. We expire producer ids by time. However, when we are deleting segments whose retention period is elapsed, we use offsets to determine whether the producer snapshot should go. This could mean that the producer id is removed even though a control message is left. This should be fine, because that message is orphaned, and will be deleted eventually once it violates retention. I don't think this is a correctness issue at all.

hachikuji replied:
Yes, I said it made me uncomfortable, not that it was necessarily a problem. I would be fine renaming the method to `lastDataOffset` as suggested above so that it's clear that it is not intended to include control records.

hachikuji replied:
We can remove the `TODO` now?

hachikuji commented on core/src/main/scala/kafka/tools/DumpLogSegments.scala:
```diff
@@ -154,10 +154,10 @@ object DumpLogSegments {
 
   private def dumpPidSnapshot(file: File): Unit = {
```
nit: rename to `dumpProducerIdSnapshot`.

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -265,11 +308,34 @@ object ProducerStateManager {
+    new Field(ProducerEpochField, Type.INT16, "Current epoch of the producer"),
+    new Field(CoordinatorEpochField, Type.INT32, "The epoch of the last transaction coordinator to send an end transaction marker"),
+    new Field(CurrentTxnFirstOffsetField, Type.INT64, "The first offset of the on-going transaction (-1 if there is none)"),
+    new Field(RecordBatchMetadataField, new ArrayOf(RecordBatchMetadataSchema), "The record metadata for up to the last 5 batches appended by the producer")
```
Related to my previous question, but I'm wondering how likely is it for us to hit a case where we need the additional entries after recovering the log. We would have to receive the produce requests, successfully write them to the log, and await enough time for a new snapshot to be written (which happens after rolling a segment and upon shutdown). All of that before the producer receives the responses. The cost of having this seems minor, but it's still useful to understand how much extra resilience it actually affords.

hachikuji reviewed (commented):
Thanks for the patch. Left a few comments. Note that I left out comments related to expiration of batches when their ultimate outcome is unknown since we discussed this in detail offline.

apurvam commented:
I created a short writeup about these changes. It goes over the assumptions made, the basic solution, and also some follow up work that needs to be done to handle corner cases:

https://docs.google.com/document/d/1EBt5rDfsvpK6mAPOOWjxa9vY0hJ0s9Jx9Wpwciy0aVo/edit?usp=sharing

Commit:
WIP

Commit:
Handled remaining corner cases:

  1. Batch expiry on the producer. In this case, we need mute the
partition and make sure all the inflight requests are fully resolved. If
they are successful, then all is well. If they are also expired or fail
due to OutOfOrderSequence, we need to reset the producer id.

  2. We now deal correctly with out of order responses to in flight
requests which may happen due to leadership changes. We keep track of
all inflight batches, ordered by sequence. Whenever there is an error,
we reduce to one inflight request by ensuring only the next batch
by sequence can be drained. When we requeue a batch, we check to make
sure that the queue remains in sequence order. When a batch completes,
it is removed from the in flight list, and the next batch by sequence
changes.

Some other changes:

 1. We never change sequence numbers unless we are certain that a batch
failed fatally on the broker (got a fatal error code), _and_ future
batches retured OutOfOrderSequence. Before the latest commit, we would
always reset sequence numbers on retry, and have other mechanisms to
ensure that a given batch got the same sequence as before on the next
drain. This is now much tighter, and needs to be in order to correctly
handle the out of order response problem.

Todo:

Add unit tests for out of order responses.
Add unit tests new cases around batch expiry (all batches expired, all
batches succeeded, future batches failed with OutOfSequence).

Commit:
Rebased onto trunk and fixed some merge conflicts

apurvam commented:
Latest updates (copied from my commit message): 

1. Batch expiry on the producer. In this case, we need mute the
partition and make sure all the inflight requests are fully resolved. If
they are successful, then all is well. If they are also expired or fail
due to OutOfOrderSequence, we need to reset the producer id.

  2. We now deal correctly with out of order responses to in flight
requests which may happen due to leadership changes. We keep track of
all inflight batches, ordered by sequence. Whenever there is an error,
we reduce to one inflight request by ensuring only the next batch
by sequence can be drained. When we requeue a batch, we check to make
sure that the queue remains in sequence order. When a batch completes,
it is removed from the in flight list, and the next batch by sequence
changes.

Some other changes:

 1. We never change sequence numbers unless we are certain that a batch
failed fatally on the broker (got a fatal error code), _and_ future
batches retured OutOfOrderSequence. Before the latest commit, we would
always reset sequence numbers on retry, and have other mechanisms to
ensure that a given batch got the same sequence as before on the next
drain. This is now much tighter, and needs to be in order to correctly
handle the out of order response problem.

Todo:

Add unit tests for out of order responses.
Add unit tests new cases around batch expiry (all batches expired, all
batches succeeded, future batches failed with OutOfSequence).

Commit:
Added unit tests for out of order responses and expiring batches

Commit:
Address PR comments on client side code

Commit:
WIP

Commit:
Revert changes where we stored the metadata for the last 5 batches. We don't need this information except in extremely rare cases, and it doesn't seem worth the cost of supporting two on disk versions of the snapshot, etc.

Commit:
Addressed a few more PR comments

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -249,6 +249,16 @@ private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime,
+            ProducerIdAndEpoch producerIdAndEpoch = new ProducerIdAndEpoch(producerId(), producerEpoch());
+            for (ProducerBatch newBatch : batches) {
+                newBatch.setProducerState(producerIdAndEpoch, sequence, isTransactional());
+                newBatch.retry = true;
```
Alternatively, we could set `retry` to be true in the constructor when `isSplitBatch` is enabled? Or is there a reason we only want to set this when idempotence is in use?

hachikuji replied:
Would this suggestion work?

hachikuji replied:
Discussed offline, but after looking again, it seems we no longer need this at all.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -332,11 +333,53 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
     }
 
+    private void maybeEnsureQueueIsOrdered(Deque<ProducerBatch> deque, ProducerBatch batch) {
+        if (transactionManager != null) {
```
nit: only two uses of this and one of them already has a check for `transactionManager` being null. Maybe we could change this to `ensureQueueIsOrdered` and add a null check for the other case.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -332,11 +333,53 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
+                    orderedBatches.add(deque.pollFirst());
+
+                log.debug("Reordered incoming batch with sequence {} for partition {}. It was placed in the queue at " +
+                        "position {}" + incomingBatch.baseSequence(), incomingBatch.topicPartition, orderedBatches.size());
```
Inadvertent string append?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -332,11 +333,53 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
         return numSplitBatches;
     }
 
+    private void maybeEnsureQueueIsOrdered(Deque<ProducerBatch> deque, ProducerBatch batch) {
```
It might be more natural to add an `insertInOrder` function (which executes the logic below modulo the first `pollFirst`) rather than inserting out of order and fixing up later.

Another thing I was wondering is whether a deque is still the right data structure given that we now require ordering. Would a priority queue give us everything we need?

apurvam replied:
I thought about this, and the deque still seems like the right one for performance reasons. We use the priority queue _just_ for in flight batches for starters, so it is smaller. Also, we use it for a variety of purposes (like reducing to 1 in flight when there are retries). And we only need to re order the deque _extremely_ rarely, so it is worth making that operation slightly more inefficient in the interests of keeping the performance for the common case.

Another thing is that I wanted to keep the behavior when idempotence is not enabled just the same, which is another reason I went this route.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -332,11 +333,53 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
+
+                ProducerBatch incomingBatch = deque.pollFirst();
+                List<ProducerBatch> orderedBatches = new ArrayList<>();
+                while (deque.peekFirst() != null && deque.peekFirst().hasSequence() && deque.peekFirst().baseSequence() < incomingBatch.baseSequence())
```
Maybe worth a comment that we depend on the invariant that assigned sequence numbers for batches on the queue are all from the same producerId.

apurvam replied:
This is a good point. If the producer id changes, that means the transaction manager will stop tracking inflight batches for the previous producer id. So I updated the code to only try reordering if there are actually in flight batches by the current producer id. 

Once we have batches by the current producer id, we know that that all batches by the previous producer id are no longer in play, because that is the only way to get out of the unresolved state and start draining new batches.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -634,7 +698,7 @@ void abortUndrainedBatches(RuntimeException reason) {
             boolean aborted = false;
             synchronized (dq) {
-                if (!batch.isClosed()) {
+                if ((transactionManager != null && !batch.hasSequence()) || (transactionManager == null && !batch.isClosed())) {
```
Hmm.. seems we should always check `!batch.isClosed()`?

apurvam replied:
This is related to your other point about the state of the records builder. Right now, `isClosed` is just `builtRecords != null`. But we do reopen the batch (ie. nullify the `builtRecords`) when we have to change the sequence numbers due to a fatal failure. That is why I had to modify this check.

hachikuji replied:
Since this function is for aborting undrained batches, maybe an alternative is to have an `isUndrained` method on `ProducerBatch` which checks if `drainedMs` is greater than 0?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -598,16 +607,28 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
             }
+            transactionManager.stopTrackingBatch(batch);
+            if (adjustSequenceNumbers)
+                transactionManager.adjustSequencesDueToFailedBatch(batch);
```
Discussed offline, but I guess the one case where it is not safe to adjust sequence numbers is when the number of retries are exhausted.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -598,16 +607,28 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
+        return batch.attempts() < this.retries &&
+                ((error.exception() instanceof RetriableException) ||
+                        (error.exception() instanceof OutOfOrderSequenceException &&
+                                transactionManager.hasProducerId(batch.producerId()) &&
```
nit: a little difficult to follow these checks. I wonder if we could wrap them in a `TransactionManager` method with a friendly name?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -67,7 +67,27 @@
 
-    private final Map<TopicPartition, Integer> sequenceNumbers;
+    // The base sequence of the next batch bound for a given partition.
+    private final Map<TopicPartition, Integer> nextSequence;
```
High level comment. There is a lot of per-partition state that we're tracking with separate collections. I wonder if this would be easier to manage with a `PartitionState` object and a single `Map<TopicPartition, PartitionState>`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,142 @@ synchronized void resetProducerId() {
+        if (!batch.hasSequence())
+            throw new IllegalStateException("Can't track batch for partition " + batch.topicPartition + " when sequence is not set.");
+        if (!inflightBatchesBySequence.containsKey(batch.topicPartition)) {
+            inflightBatchesBySequence.put(batch.topicPartition, new PriorityQueue<>(5, new Comparator<ProducerBatch>() {
```
nit: maybe we could add a private static extension of PriorityQueue which sets the comparator?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -279,6 +279,10 @@ public void abort() {
     }
 
+    public void unsetProducerState() {
+        builtRecords = null;
```
What should the state of the builder be for this? Maybe we should verify at least that it wasn't aborted?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,142 @@ synchronized void resetProducerId() {
+                (lastAckedSequence.containsKey(topicPartition) && (sequence - lastAckedSequence.get(topicPartition) == 1));
+    }
+
+    synchronized  void setNextSequence(TopicPartition topicPartition, int sequence) {
```
nit: this could be `private`? Also, there's an unneeded space after `synchronized`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,142 @@ synchronized void resetProducerId() {
+        return inflightBatchesBySequence.get(topicPartition).peek();
+    }
+
+    synchronized void stopTrackingBatch(ProducerBatch batch) {
```
nit: maybe a better name is something like `addInFlightBatch`?

hachikuji reviewed (commented):
Thanks for the updates. Mostly finished with the changes on the client side.

Commit:
Address review comments for the broker side chanegs. Added tests for the new duplicate detection logic.

apurvam commented:
One new problem we have to tackle with the new approach to duplicates. When sequnce numbers wrap around, it is possible to detect spurious duplicates. There doesn't seem to be an easy solution for this which is elegant, un less we put some bound on how old a duplicate can be. This also is worth discussing.

Commit:
Handle duplicates during wraparound

Commit:
Two changes:

  1. A solution for detecting duplicates when sequence numbers
wraparound. This is essentially the same as the solutior for detecing
out of order sequences with wraparound, but both are limited.
  2. Correctly handle duplicate sequence exceptions on the client.

Commit:
Addressed more PR comments

apurvam commented:
retest this please

apurvam commented:
Started system tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1071/

Commit:
Address a few more minor PR comments

Commit:
Revert changes to log4j.properties

Commit:
Delete unnecessary trailing whitespace

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -310,6 +310,7 @@ public void reenqueue(ProducerBatch batch, long now) {
         Deque<ProducerBatch> deque = getOrCreateDeque(batch.topicPartition);
         synchronized (deque) {
             deque.addFirst(batch);
+            maybeEnsureQueueIsOrdered(deque, batch);
```
Mentioned previously, but I think it would be more intuitive to insert in order rather than fixing up the collection after the insertion.

hachikuji replied:
Would like to hear your thoughts on my previous comment as well:
>Another thing I was wondering is whether a deque is still the right data structure given that we now require ordering. Would a priority queue give us everything we need?

hachikuji commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -279,6 +279,13 @@ public void abort() {
         aborted = true;
     }
 
+    public void unsetProducerState() {
```
Hmm.. This doesn't actually unset the producer state. Maybe `reopen` would be a better name?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,144 @@ synchronized void resetProducerId() {
+        Integer currentLastAckedSequence = lastAckedSequence.get(topicPartition);
+        if (currentLastAckedSequence == null)
+            return -1;
+        return lastAckedSequence.get(topicPartition);
```
Can't we return `currentLastAckedSequence`?

apurvam replied:
I chose to return -1 when there is no last ack'd sequence because it makes the logic elsewhere a bit cleaner, for instance is `maybeUpdateLastAckdSequence` we don't need a null check with the current setup.

hachikuji replied:
That is fine, I am just questioning why we don't use the variable `currentLastAckedSequence` instead of calling `lastAckedSequence.get(topicPartition)` a second time.

apurvam replied:
ack. Will update.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,144 @@ synchronized void resetProducerId() {
+    }
+
+    synchronized void removeInFlightBatch(ProducerBatch batch) {
+        if (inflightBatchesBySequence.containsKey(batch.topicPartition))
```
nit: Of the micro-optimization variety, but I think it's a little nicer pattern to do a `get` followed by a `null` check rather than two hash lookups.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -332,11 +333,64 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
+            // id must have changed and the batches being re enqueued are from the old producer id. In this case
+            // we don't try to ensure ordering amongst them. They will eventually fail with an OutOfOrderSequence,
+            // or they will succeed.
+            if (transactionManager.nextBatchBySequence(batch.topicPartition) != null &&
```
nit: annoying that we make two calls to `nextBatchBySequence` here. Maybe we can use a local variable?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,144 @@ synchronized void resetProducerId() {
+
+
+    synchronized ProducerBatch nextBatchBySequence(TopicPartition topicPartition) {
+        return inflightBatchesBySequence.get(topicPartition).peek();
```
Why do we not need the null check here, but we do in `removeInFlightBatch`? Should this be documented at least?

apurvam replied:
Good catch. Actually, we do need the null check here. The reason it exists in both places is that when we reset producer the producer id, if there is a queued batch in retry from the previous producer id, we would hit an NPE when trying to drain it after the reset. I have fixed it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,144 @@ synchronized void resetProducerId() {
+        return inflightBatchesBySequence.containsKey(topicPartition) && !inflightBatchesBySequence.get(topicPartition).isEmpty();
+    }
+
+    synchronized boolean partitionHasUnresolvedSequence(TopicPartition topicPartition) {
```
nit: since the argument is a partition, we could probably drop the `partition` prefix from the name.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,144 @@ synchronized void resetProducerId() {
+    }
+
+    synchronized boolean isNextSequence(TopicPartition topicPartition, int sequence) {
+        return (!lastAckedSequence.containsKey(topicPartition) && sequence == 0) ||
```
nit: similar to above, but there are three hash lookups here in the worst case when only one is needed.

apurvam replied:
good suggestion. By returning `-1` from `lastAckedSequence` if there isn't a last acked sequence, we can reduce this to `sequence - lastAckedSequence(topicPartition) == 1`. Much simpler!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -563,21 +574,24 @@ private void reenqueueBatch(ProducerBatch batch, long currentTimeMs) {
-                    transactionManager.sequenceNumber(batch.topicPartition));
+        if (transactionManager != null) {
+            if (transactionManager.hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch())) {
+                transactionManager.maybeUpdateLastAckedSequence(batch.topicPartition, batch.baseSequence() + batch.recordCount - 1);
```
nit: if you use a local variable for `lastSequence`, then you can also use it in the log message below.

apurvam replied:
Actually, the log message should contain the actual lastAckedSequence since the update is conditional: with out of order responses, it is possible that we might move the sequence backward, so the `maybeUpadteLastAckedSequence` guards against that.

tedyu replied:
Pardon.

In the current code base, I don't see maybeUpadteLastAckedSequence.
What's the method called now ?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -537,9 +547,10 @@ else if (error == Errors.CLUSTER_AUTHORIZATION_FAILED)
-                failBatch(batch, response, exception);
-                this.sensors.recordErrors(batch.topicPartition.topic(), batch.recordCount);
+                // tell the user the result of their request. We only adjust sequence numbers if the batch didn't exhaust
+                // its retries -- if it did, we don't know the whether the sequence number was accepted or not, and
```
typo: drop "the" before "whether"

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -48,33 +48,80 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
+    if (this.producerEpoch != producerEpoch) {
+      batchMetadata.clear()
+      this.producerEpoch = producerEpoch
+      return true
```
nit: explicit returns are frowned upon in scala. Maybe we can drop the `return` and add an `else` branch? Same for `duplicateOf` below.

hachikuji reviewed (commented):
A few more (smaller) comments.

apurvam commented:
retest this please

apurvam commented:
Some tests here are definitely leaking threads. Trying to reproduce by running tests in sequence locally.

Commit:
Let the producerIdEntry have a concurrent structure for batch metadata. This makes accesses from the cleaner threads (log cleaner, producer id expiration, etc.) safe

Commit:
Bunch of changes:

  1. Fixed a potentioal NPE in ProducerStateManager.writeSnapshot where
we could have no batch metadata in memory (for instance, if only a
control batch was written), and yet write a snapshot. This was causing
tests to fail because an NPE was raised during shutdown, leaking
threads.
  2. Addressed the remaining PR comments.

apurvam reviewed (commented):
@hachikuji I addressed most of your comments.

I also found the cause of the thread leaks: in the new scheme, the in memory `batchMetadata` only caches data batches. However, if an instance of the `ProducerStateManager` only got a control record for a producer, we would have a `ProducerIdEntry` with empty batch metadata. When we go to write a snapshot, the new code would NPE, causing threads to leak in shutdown, and causing _other_ tests to fail.

Commit:
Rename ProducerIdEntry.lastOffset to ProducerIdEntry.lastDataOffset everywhere

apurvam commented:
> However, if an instance of the ProducerStateManager only got a control record for a producer, we would have a ProducerIdEntry with empty batch metadata. When we go to write a snapshot, the new code would NPE, causing threads to leak in shutdown, and causing other tests to fail.

The specific test causing the problem was `TransactionsTest.testFencingOnSend`. In the current patch, we drop the batch metadata when be bump the epoch because the previous sequence numbers no longer apply and shouldn't be checked. In this test, we write a snapshot immediately after after the fence, which triggers the NPE.

Commit:
Address more PR comments

Commit:
Fix previous commit which introduced a bug in RecordAccmulator.drain.
If a batch has a sequence, but there are no recorded inflight requests
to that partition, it means that the producer id has changed, and the
batch under consideration was from a previous producerId. It should
always be drained (if it returns again, it will fail fatally).

The code before this patch would ensure that it would _never_ drain.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -365,28 +389,148 @@ synchronized void resetProducerId() {
+                // The partition has been fully drained. At this point, the last ack'd sequence should be once less than
+                // next sequence destined for the partition. If so, the partition is fully resolved. If not, we should
+                // reset the sequence number if necessary.
+                if (lastAckedSequence(topicPartition) == sequenceNumber(topicPartition) - 1) {
```
I think we could use `isNextSequence` here.

Commit:
PR comment: return currentLastSequence instead of doing another map lookup

Commit:
Remove the concurrent queue from ProducerIdEntry and instead return just the lastSeq per producer to the LogCleaner, hence simpliying the concurrency model

Commit:
Address other minor comments

Commit:
Correctly transition to a fatal error state when batches in retry are expired during a transaction

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -467,6 +470,560 @@ public void testClusterAuthorizationExceptionInInitProducerIdRequest() throws Ex
+        assertEquals(0, transactionManager.sequenceNumber(tp0).longValue());
+
+        // Send first ProduceRequest
+        Future<RecordMetadata> request1 = accumulator.append(tp0, time.milliseconds(), "key".getBytes(), "value".getBytes(), null, null, MAX_BLOCK_TIMEOUT).future;
```
I think we should be asserting on the result of the send futures in these test cases.

apurvam replied:
Just added the assertions for test cases in which they were missing.

Commit:
Added assertions on the send futures for some of the newly added test cases in SenderTest

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -467,6 +470,571 @@ public void testClusterAuthorizationExceptionInInitProducerIdRequest() throws Ex
+
+        assertEquals(1, client.inFlightRequestCount());
+        assertEquals(0, transactionManager.lastAckedSequence(tp0));
+        assertNotNull(request1.get());
```
Can we assert the offset? This ensures that the requests have not been reordered. Also, the `isDone` check should probably come first

apurvam replied:
Hmm. The `lastAckdSequence` checks already ensure that the responses are coming in the right order. We also check the sequence of the outgoing batch matches what we expect when the requests are sent, which is really what we want to check in these tests.

Commit:
check the response offsets in all the tests

Commit:
Fix checkstyle failures

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -331,12 +334,68 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
+        // we don't try to ensure ordering amongst them. They will eventually fail with an OutOfOrderSequence,
+        // or they will succeed.
+        if (transactionManager.nextBatchBySequence(batch.topicPartition) != null &&
+                batch.baseSequence() != transactionManager.nextBatchBySequence(batch.topicPartition).baseSequence()) {
```
nit: here we use sequence number equality to tell when two batches match. Below in `drain` we use `equals`: https://github.com/apache/kafka/pull/3743/files#diff-e67ae6379c50ac1d9c58cefaf71346ceR540. Unless there is a good reason, maybe we should be consistent?

apurvam replied:
ack. I prefer to just use sequence number equality. Will update the `drain` method.

Commit:
Address further PR comments

apurvam commented:
The system tests all passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-09-11--001.1505200401--apurvam--KAFKA-5494-increase-max-in-flight-for-idempotent-producer--e2bd0bc/report.html

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -228,6 +231,12 @@ void run(long now) {
         Cluster cluster = metadata.fetch();
+
+
+        if (transactionManager != null && transactionManager.shouldResetProducerStateAfterResolvingSequences()) {
```
Seems like there's no longer any reason for this logic to be in `sendProducerData`. Maybe we may as well move it into `run(now)` with the rest of the other periodic transaction/idempotent checks?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -467,6 +469,583 @@ public void testClusterAuthorizationExceptionInInitProducerIdRequest() throws Ex
+        sender.run(time.milliseconds());  // send second request
+        sendIdempotentProducerResponse(1, tp0, Errors.OUT_OF_ORDER_SEQUENCE_NUMBER, 1);
+        sender.run(time.milliseconds()); // receive second response, the third request shouldn't be sent since we are in an unresolved state.
+        assertTrue(request2.isDone());
```
Should we assert the failure exception?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -467,6 +469,583 @@ public void testClusterAuthorizationExceptionInInitProducerIdRequest() throws Ex
+
+        assertEquals(2, client.inFlightRequestCount());
+
+        sendIdempotentProducerResponse(0, tp0, Errors.NOT_LEADER_FOR_PARTITION, -1);
```
nit: maybe TIMEOUT_ERROR would be better for this case since NOT_LEADER_FOR_PARTITION means the record definitely wasn't written. 

By the way, doesn't this expose a weakness in our error checking? We know for sure that the write never succeeded even though we had to expire it.

apurvam replied:
Yes. I think this was brought up in the KIP-91 discussion as well: we could optimize for error codes which definitively indicate that the batch wasn't written, but we don't do that today. 

This means we could get into situations where we reset the producer id where we don't strictly need to. I filed https://issues.apache.org/jira/browse/KAFKA-5897 to track this.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -467,6 +469,583 @@ public void testClusterAuthorizationExceptionInInitProducerIdRequest() throws Ex
 
+
+    @Test
+    public void testIdempotenceWithMultipleInflights() throws Exception {
```
I was looking for a test case which exercised the sequence number adjustment logic, but couldn't find one. Specifically a test case which caused the sequence number of a batch to change.

apurvam replied:
Good catch. There wasn't a test, and there was a bug in that logic. Particularly, when a batch fails fatally, the future in flight batches are guaranteed to fail with an `OutOfOrderSequenceException`. We retry these exceptions as long as the batch in question is not the next-in-line batch. However, when we reset sequence numbers, the first future batch does become the next-in-line batch, and it is not retried when it fails with --the expected-- `OutOfOrderSequenceException`. 

I have have the test case and the fix for the bug in an upcoming commit.

hachikuji reviewed (commented):
A few more comments. Getting close!

Commit:
Address PR comments.

One notable update: We didn't have a test for reassiging sequence
numbers, and we had a bug in that logic. Essentially a newly reassigned
batch would be fail with an out of order sequence exception because it
would be the next in line batch after its sequence was adjusted.

This is fixed by tracking whether the sequence of a batch has been
adjusted since the last drain, in which case the out of order sequence
exception for it is always retried. If it fails again, it will be fatal.

Commit:
Fix up build failures after merge

Commit:
Address more PR comments:

  1. Move the 'reopen' flag from MemoryRecordsBuilder to ProducerBatch
  2. Move the logic to resolve sequence number state to the top of the
run loop.
  3. More canonical scala in Log.scala.

Commit:
Address final PR comment

hachikuji reviewed (approved):
LGTM. I'll merge after the builds complete.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -331,12 +334,71 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
+
+        // If there are no inflight batches being tracked by the transaction manager, it means that the producer
+        // id must have changed and the batches being re enqueued are from the old producer id. In this case
+        // we don't try to ensure ordering amongst them. They will eventually fail with an OutOfOrderSequence,
```
OutOfOrderSequence -> OutOfOrderSequenceException

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -362,28 +386,157 @@ synchronized void resetProducerId() {
+                }
+            }));
+        }
+        inflightBatchesBySequence.get(batch.topicPartition).offer(batch);
```
Shouldn't we add a check for the number of batches stored in the PriorityQueue ?
What if more than 5 batches are stored ?

luigiberrettini commented:
I saw that the [Sender](https://github.com/apache/kafka/blob/2.6.0/clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java#L602) checks for a `Errors.DUPLICATE_SEQUENCE_NUMBER` but I was not able to find where this error is triggered on the server side.

It seems to me that duplicate detection relies on checking if the sequence number is more than `lastPersistedSeq + 1`.
If this is the case:
 - why storing the metadata for the last batches and not just relying on the sequence number of the last message persisted in the log?
 - why limiting `max.in.flight.requests.per.connection` to a maximun value of 5 if duplicates are still detected when metadata is not found (and therefore with any number of max in flights)?
